app/api/v1/auth/index.js

```diff
@@ -82,16 +83,51 @@ router.post('/register', (req, res, next) => {
     return next(new error.BadRequest('Password must be at least 10 characters long'));
   }
 
-  // TODO account activation via email
+
   const newUser = User.sanitize(req.body.user);
   newUser.state = 'ACTIVE';
   User.generateHash(req.body.user.password).then((hash) => {
     newUser.hash = hash;
-    return User.create(newUser);
-  }).then((user) => {
-    log.info('user authentication (registration)', { request_id: req.id, user_uuid: user.uuid });
-    res.json({ error: null, user: user.getPublicProfile() });
-    Activity.accountCreated(user.uuid);
+    // require that we create a user succesfully and send email succesfully
+    db.transaction({
+      isolationLevel: Sequelize.Transaction.ISOLATION_LEVELS.REPEATABLE_READ,
+    }, (transaction) => Promise.all([
+      User.create(newUser).then((user) => {
+        User.generateAccessCode().then((code) => {
+          user.accessCode = code;
+          email.sendEmailVerification(user.email, user.firstName, code).then(() => {
+            // save the access code if email was sent
+            user.save();
+          }).catch(() => {
+            throw new error.BadRequest(`Something went wrong with sending email verification to ${user.email}`);
+          });
+        });
+        return user;
+      }),
+    ])).then((transactRes) => {
+      const user = transactRes[0];
+      log.info('user authentication (registration)', { request_id: req.id, user_uuid: user.uuid });
+      res.json({ error: null, user: user.getPublicProfile() });
+      Activity.accountCreated(user.uuid);
+    }).catch((err) => next(err));
+  }).catch(next);
+});
+
+/**
+ * Emails the user a email verification link
+ */
+// TODO: RATE LIMIT THIS
+router.post('/emailVerification', (req, res, next) => {
+  if (!req.body.email) return next(new error.BadRequest('Email must be provided!'));
+  User.findByEmail(req.body.email).then((user) => {
+    if (!user) return next(new error.BadRequest('Invalid email'));
```

Comment:
Throw errors inside Promises so the `catch` block passes it to `next` (I've run into some weird errors otherwise).

Reply:
Makes a lot of sense actually, returning `next` inside a passed in function is kinda wierd because `next` is associated with the enclosing function